src/diff.c

```diff
@@ -2004,18 +2004,92 @@ lyd_diff_change_op(struct lyd_node *node, enum lyd_diff_op op)
     }
 }
 
+/**
+ * @brief In user-ordered lists, certain operations on sibling nodes can result in logically identical changes.
+ *        However, applying the first change may cause the second one to fail.
+ *        To prevent this, the affected node is unlinked.
+ *
+ * @param[in,out] diff The node whose metadata has been modified.
+ * @param[in] mod  The YANG module associated with the metadata.
+ * @return LY_ERR value.
+ */
+static LY_ERR
+lyd_diff_find_and_unlink_cyclic_nodes(struct lyd_node **diff, const struct lys_module *mod)
+{
+    LY_ERR ret = LY_SUCCESS;
+    struct lyd_meta *metas[2];
+    struct lyd_node *diff_iter = *diff;
+    char *buff1 = NULL, *buff2 = NULL;
+    const char *name = NULL, *name_iter = NULL;
+    size_t bufflen1 = 0, buffused1 = 0;
+    size_t bufflen2 = 0, buffused2 = 0;
+
+    /* Itereate throught previous nodes and look for logically identical changes */
+    while (diff_iter->prev->next != NULL) {
+        diff_iter = diff_iter->prev;
+        metas[0] = lyd_find_meta((*diff)->meta, mod, "key");
+        metas[1] = lyd_find_meta(diff_iter->meta, mod, "orig-key");
+
+        /* If keys don't match, skip - not a candidate for cyclic change */
+        if (strcmp(lyd_get_meta_value(metas[0]), lyd_get_meta_value(metas[1]))) {
+            continue;
+        }
+
+        metas[0] = lyd_find_meta((*diff)->meta, mod, "orig-key");
+        metas[1] = lyd_find_meta(diff_iter->meta, mod, "key");
+
+        /* Store string values of metadata to compare later */
+        name = lyd_get_meta_value(metas[0]);
+        name_iter = lyd_get_meta_value(metas[1]);
+
+        if ((ret = lyd_path_list_predicate(*diff, &buff1, &bufflen1, &buffused1, 0))) {
+            return ret;
+        }
+
+        if ((ret = lyd_path_list_predicate(diff_iter, &buff2, &bufflen2, &buffused2, 0))) {
+            return ret;
+        }
+
+        /* Compare path predicates with metadata - check if this is a reversed operation */
+        if (!strcmp(buff1, name_iter) && !strcmp(buff2, name)) {
+
+            /* Found a cyclic change - remove and free the node */
+            if ((ret = lyd_unlink_tree(*diff))) {
+                return ret;
+            }
+
+            lyd_free_all(*diff);
+            free(buff1);
+            free(buff2);
+            *diff = NULL;
+            return ret;
+        }
+
+        free(buff1);
+        buff1 = NULL;
+        bufflen1 = 0;
+        buffused1 = 0;
+        free(buff2);
+        buff2 = NULL;
+        bufflen2 = 0;
+        buffused2 = 0;
```

Comment:
Seems like a good candidate for a `cleanup` label where everything is freed. In between iterations it should be enough to zero `buffused`.